src/bin/main.rs

```diff
@@ -101,40 +103,71 @@ fn run (args: Args) -> Result<(), Error> {
              0x89, 0xc8,
              0x5d,
              0xc3])?;
+
     // main:
     // 55	push   %rbp
     // 48 89 e5	mov    %rsp,%rbp
+    // 48 83 ec 10	sub    $0x10,%rsp
+    // c7 45 fc 00 00 00 00	movl   $0x0,-0x4(%rbp)
     // b8 00 00 00 00	mov    $0x0,%eax
-    // e8 00 00 00 00   callq  0x0 <deadbeef>
+    // e8 00 00 00 00	callq  0x16 <deadbeef>
+    // 48 8d 3d 00 00 00 00	lea    0x0(%rip),%rdi        # 0x1d <main+29> will be: "deadbeef: 0x%x - %d\n"
+    // 48 8b 0d 00 00 00 00	mov    0x0(%rip),%rcx        # 0x24 <main+36>
+    // 8b 11	mov    (%rcx),%edx
     // 89 c6	mov    %eax,%esi
-    // 48 8d 3d 00 00 00 00 lea    0x0(%rip),%rdi # will be: deadbeef: 0x%x\n
-    // b8 00 00 00 00	mov    $0x0,%eax
-    // e8 00 00 00 00	callq  0x3f <main+33>  # printf
-    // b8 00 00 00 00	mov    $0x0,%eax
+    // b0 00	mov    $0x0,%al
+    // e8 00 00 00 00	callq  0x2f <main+47> # printf
+    // 31 d2	xor    %edx,%edx
+    // 89 45 f8	mov    %eax,-0x8(%rbp)
+    // 89 d0	mov    %edx,%eax
+    // 48 83 c4 10	add    $0x10,%rsp
     // 5d	pop    %rbp
     // c3	retq
     obj.define("main",
-        vec![0x55,
+        vec![
+             0x55,
              0x48, 0x89, 0xe5,
+             0x48, 0x83, 0xec, 0x10,
+             0xc7, 0x45, 0xfc, 0x00, 0x00, 0x00, 0x00,
              0xb8, 0x00, 0x00, 0x00, 0x00,
              0xe8, 0x00, 0x00, 0x00, 0x00,
-             0x89, 0xc6,
              0x48, 0x8d, 0x3d, 0x00, 0x00, 0x00, 0x00,
-             0xb8, 0x00, 0x00, 0x00, 0x00,
+             0x48, 0x8b, 0x0d, 0x00, 0x00, 0x00, 0x00,
+             0x8b, 0x11,
+             0x89, 0xc6,
+             0xb0, 0x00,
              0xe8, 0x00, 0x00, 0x00, 0x00,
-             0xb8, 0x00, 0x00, 0x00, 0x00,
+             0x31, 0xd2,
+             0x89, 0x45, 0xf8,
+             0x89, 0xd0,
+             0x48, 0x83, 0xc4, 0x10,
              0x5d,
-             0xc3])?;
-    obj.define("str.1", b"deadbeef: 0x%x\n\0".to_vec())?;
+             0xc3,
+        ])?;
+    // define static data
+    obj.define("str.1", b"deadbeef: 0x%x - 0x%x\n\0".to_vec())?;
+    obj.define("STATIC",     [0xbe, 0xba, 0xfe, 0xca].to_vec())?;
+    // FIXME: @sunfishcode, et. al: I think this is a hack, but I'm not sure what the best
+    // way to tell the artifact layer that this is a reference, and gets its value from another symbol
+    // we may need a new DeclType::Ref or something? Needs discussion
+    obj.define("STATIC_REF", vec![0; 8])?;
```

Comment:
Are you concerned that filling it until with zero bytes which will be overwritten by the relocation is a hack? If so, I don't think it's a problem.

This is a special case of a more general need. Think of a global variable with type array of struct containing a mix of references and other stuff. We'll just have one big initializer with zeros where the relocations go.

Reply:
Ok, sounds good to me! As an optimization, might be nice to introduce an allocationless define_zeroes or something like that, and the backend would write out zeroes for you. Though this might not be worth the trouble or complexity for a perhaps somewhat small usecase? Although, once proper bss support gets added, might have to add something like this anyway

Reply:
Fwiw, cranelift-module already has [`define_zeroinit`](https://docs.rs/cranelift-module/0.21.1/cranelift_module/struct.DataContext.html#method.define_zeroinit) for setting the initialization to zeros (independently of any relocations on top of that), in anticipation of exactly this :-).